Add support for common Auth headers

AntoniKania opened the pull request:
<!-- Please describe your pull request here. -->

## References
PR resolves #24 

<!-- References to relevant GitHub issues and pull requests, esp. upstream and downstream changes -->
## Description
Following methods were added `WithAuthToken`, `WithBearerToken`, `WithDigestAuth`, all of them are taking `tokenMatcher BasicParamMatcher` as an parameter, e.g.: `WithBearerToken(Contains("123").And(StartsWith("token"))).`, more examples can be seen in [test scenarios](https://github.com/AntoniKania/go-wiremock/blob/add-support-for-common-auth-headers/client_test.go#L77). Additionally a new `StartsWith` method that uses a `matches` `ParamMatchingStrategy` underneath was added for creating `StringValueMatcher`.

### Draft PR notes
I faced some issues when it comes to the [`AddPrefixToMatcher`](https://github.com/AntoniKania/go-wiremock/blob/add-support-for-common-auth-headers/string_value_matcher.go#L59) method implementation - the method is intended to add a Auth method (e.g.: `Bearer `) to the matcher that was passed to the `WithBearerToken` where only matcher intended to match the token itself was passed. The goal here is to generate a matcher that will generate a [whole assertion](https://github.com/AntoniKania/go-wiremock/blob/add-support-for-common-auth-headers/testdata/expected-template-bearer-auth-equalTo.json#L12) - e.g. for a `EqualTo`, the `Bearer ` part needs to be added to the original `tokenMatcher` passed by the user (for EqualTo it works fine now). The main problem are regex's passed, e.g., to the `Matching` method. Right now, I'm trying to modify the past regex so it would work correctly with the Auth method prefix placed in the header, but I really think that this implementation is not covering more complex regex's (I have only covered cases which I could think of). Another thing is that I don't know how to properly handle cases when the new regex creation will create malformed regex. For now, I'm printing the message (which I bet can't stay) and returning the original matcher, I believe that returning error here is not the best idea. I would really appreciate any resources pointing out how to properly resolve this problem :) 
## Submitter checklist

- [x] Recommended: Join [WireMock Slack](https://slack.wiremock.org/) to get any help in `#help-contributing` or a project-specific channel like `#wiremock-java`
- [x] The PR request is well described and justified, including the body and the references
- [x] The PR title represents the desired changelog entry
- [x] The repository's code style is followed (see the contributing guide)
- [x] Test coverage that demonstrates that the change works as expected
- [x] For new features, there's necessary documentation in this pull request or in a subsequent PR to [wiremock.org](https://github.com/wiremock/wiremock.org)

<!--
Put an `x` into the [ ] to show you have filled the information.
The template comes from https://github.com/wiremock/.github/blob/main/.github/pull_request_template.md 
You can override it by creating .github/pull_request_template.md in your own repository 
-->

AntoniKania commented:
I have done some research, and it looks like the current implementation of [AddPrefixToMatcher](https://github.com/AntoniKania/go-wiremock/blob/add-support-for-common-auth-headers/string_value_matcher.go#L59) will handle all the cases, as the regex (matching token alone) provided by the user will have to match the whole token expression, so adding this prefix to the regex should never break anything, nor produce a malformed regex.

walkerus reviewed (changes requested):
Looks good. There is one small note

AntoniKania commented:
Marking as ready for review :) Few notes from my site:
- I've changed the addPrefixToMatcher method to be unexported, as I don't think there is any value in making it available for end users, and can only cause confusion.
- I've implemented the [TODO statement](https://github.com/wiremock/go-wiremock/pull/27/commits/92b924be0b1eefb56d118b9f8e24ce652bdb531a#diff-f07990bfb3d93719c0869289d9aad3edf806b6ef9108b24961bfdb02fe9440cdR58) that was added previously, although I'm not really sure if usage of this method makes sense in those cases.

oleg-nenashev commented:
@AntoniKania @walkerus I suggest doing a follow-up pull request with some documentation on Authentication capabilities in README or additional documentation strings

AntoniKania commented:
Sure, I can add that :)

oleg-nenashev commented:
Thanks!

On Fri, Sep 1, 2023 at 1:35 PM Antoni Kania ***@***.***>
wrote:

> Sure, I can add that :)
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/wiremock/go-wiremock/pull/27#issuecomment-1702608484>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAW4RIDMC7OWZYB2COLBBMDXYHCB7ANCNFSM6AAAAAA4ARKOZ4>
> .
> You are receiving this because your review was requested.Message ID:
> ***@***.***>
>
